src/inference/graph_inference.cpp

```diff
@@ -920,60 +933,94 @@ class GraphNodeAssignment : public Nice2Assignment {
+            PerformPermutationOptimization(inf_vars, fweights, candidate_inf_labels, &best_assignments, &best_score);
+            current_num_permutation++;
           }
-
-          bool is_assignment_valid = true;
-          for (size_t z = 0; z < inf_vars.size(); ++z) {
-            if (HasDuplicationConflictsAtNode(inf_vars[z]) ||
-                !fweights.label_checker_.IsLabelValid(assignments_[inf_vars[z]].label)) {
-              is_assignment_valid = false;
-              break;
+        } else {
+          std::vector<size_t> count;
+          size_t current_num_permutations = 0;
+          count.assign(candidate_inf_labels.size(), 0);
+          for (size_t i = 1; i < candidate_inf_labels.size() && current_num_permutations < FLAGS_permutations_beam_size;) {
+            if (count[i] < i) {
+              int swap = i % 2 * count[i];
+              int tmp = candidate_inf_labels[swap];
+              candidate_inf_labels[swap] = candidate_inf_labels[i];
+              candidate_inf_labels[i] = tmp;
+              PerformPermutationOptimization(inf_vars, fweights, candidate_inf_labels, &best_assignments, &best_score);
+              current_num_permutations++;
+              count[i]++;
+              i = 1;
+            } else {
+              count[i++] = 0;
             }
           }
-          if (is_assignment_valid == false) {
-            continue;
-          }
-          double score = 0;
-          std::vector<int> assignment;
-          assignment.assign(inf_vars.size(), 0);
-          for (size_t z = 0; z < inf_vars.size(); ++z) {
-            score += GetNodeScore(fweights, inf_vars[z]);
-            assignment[z] = assignments_[inf_vars[z]].label;
-          }
-          if (score > best_score) {
-            best_assignments = assignment;
-            best_score = score;
-          }
         }
+        giv_labels = giv_labels_copy;
       }
       for (size_t j = 0; j < inf_vars.size(); ++j) {
         assignments_[inf_vars[j]].label = best_assignments[j];
       }
     }
   }
 
+  bool IsAssignmentValid(const std::vector<int>& inf_vars, const std::vector<int>& candidate_inf_labels,const GraphInference& fweights) const {
+    bool is_assignment_valid = true;
+    for (size_t z = 0; z < inf_vars.size(); ++z) {
+      if (HasDuplicationConflictsAtNode(inf_vars[z]) ||
+          !fweights.label_checker_.IsLabelValid(candidate_inf_labels[z])) {
+        return false;
+      }
+    }
+    return is_assignment_valid;
+  }
+
+  void PerformPermutationOptimization(const std::vector<int>& inf_vars,
+                                      const GraphInference& fweights,
+                                      const std::vector<int>& candidate_inf_labels,
+                                      std::vector<int>* best_assignments,
+                                      double* best_score) {
+    if (IsAssignmentValid(inf_vars, candidate_inf_labels, fweights)) {
```

Comment:
Yes, and also I deleted the IsAssignmentValid function since I moved the check if the labels are valid before I start checking the various permutations.